File: apps/storefront/lib/answer-link.ts
/**
 * The Answer Link's page, as rules rather than as markup (#312, ADR 0044).
 *
 * An Answer Link opens ONE Ticket's Ticket Questions for whoever the buyer
 * forwarded it to, and shows nothing else about the purchase. That last part is
 * a backend property — service.AnswerLinkView never sends the buyer, the price,
 * the Tax ID or the confirmation reference — and this file is the reason it
 * stays one: NOTHING HERE INVENTS A FIELD. It renders what arrived, and what
 * arrives is only ever those three things.
 *
 * Pure and dependency-free — no React, no i18n runtime — so the rules are
 * directly unit-testable, exactly as the staff app's lib/ticket-answers.ts is.
 * It returns SHAPES and never a sentence.
 *
 * WHAT IS DATA AND WHAT IS COPY. A Ticket Question's label and an Option's label
 * are read AS COINED in every Locale, like a Custom Tag (ADR 0027): the
 * Organization wrote them, and translating "T-shirt size" into the reader's
 * language would be the platform putting words in their mouth. Only the page's
 * own chrome follows the Locale.
 */

/** The seven shapes a Ticket Question takes. The kind is frozen once answered. */
export type QuestionKind =
  | "short_text"
  | "long_text"
  | "single_choice"
  | "multi_choice"
  | "number"
  | "date"
  | "checkbox";

/** One Option a choice question offers. */
export type QuestionOption = {
  id: string;
  /** Coined by the Organization; drawn as written in both languages. */
  label: string;
  sort_order: number;
  /** Retired: gone from every new list, kept on the Tickets that chose it. */
  retired: boolean;
};

/** One Ticket Question, as the public Answer Link payload carries it. */
export type Question = {
  id: string;
  label: string;
  kind: QuestionKind;
  /**
   * Required means OUTSTANDING, never BLOCKING (ADR 0044). Nothing on this page
   * is refused for want of an Answer and no control is disabled by it; the flag
   * exists so the page can say which ones the Organization is waiting on.
   */
  required: boolean;
  sort_order: number;
  retired: boolean;
  options: QuestionOption[];
};

/** One Option a choice Answer chose. */
export type AnswerOption = {
  option_id: string;
  /** THE SNAPSHOT: the words this Option showed when it was chosen. */
  label: string;
  /** What the same Option reads now. Differs from `label` after a rename. */
  current_label: string;
  retired: boolean;
};

/**
 * One Ticket's Answer to one Ticket Question.
 *
 * EXACTLY ONE OF THE FIRST FOUR IS NON-NULL, decided by the question's kind.
 * `checked: false` is an Answer — somebody read it and said no — while a null
 * answer is the absence of one.
 */
export type Answer = {
  text: string | null;
  /**
   * A decimal STRING and not a number, all the way from the NUMERIC column. A
   * JSON number would be an IEEE double by the time it reached here.
   */
  number: string | null;
  /** A calendar date, YYYY-MM-DD. Never an instant, so nothing shifts it. */
  date: string | null;
  checked: boolean | null;
  options: AnswerOption[];
  updated_at: string;
};

/** One question paired with this Ticket's reply, or null for no reply. */
export type QuestionAnswer = {
  question: Question;
  answer: Answer | null;
};

/**
 * What an Answer Link opens: THREE FIELDS AND NO MORE.
 *
 * This type is the Storefront's half of ADR 0044's disclosure rule, and it is
 * written narrow ON PURPOSE. There is no buyer here, no price, no Tax ID, no
 * Sale Confirmation reference, no sibling Ticket and not even a ticket id — the
 * form posts back with the TOKEN. A link forwarded into a group chat should tell
 * the group nothing about who paid.
 *
 * Anybody widening this is reversing that decision, and the backend will not
 * feed it: the API sends exactly these, and there is an integration test that
 * reads the raw response body and fails if a fourth appears.
 */
export type AnswerLinkView = {
  event_name: string;
  ticket_type_name: string;
  questions: QuestionAnswer[];
};

/**
 * The body a PUT to one Answer takes: exactly the field its question's kind
 * uses, plus the token that says which Ticket is answering.
 */
export type AnswerBody = {
  text?: string;
  number?: string;
  date?: string;
  checked?: boolean;
  option_ids?: string[];
};

/**
 * The `errors.answerLink` catalog key for each way a link can fail to open.
 *
 * Mapped here rather than left to the API's own sentence because these three are
 * the whole of what this page can go wrong with, and because the recovery
 * differs: an expired link has no replacement and a broken one might, so
 * offering the same words for both would send half the readers back to the buyer
 * for nothing.
 */
export const ANSWER_LINK_FAILURE_KEYS = {
  ANSWER_LINK_INVALID: "invalid",
  ANSWER_LINK_EXPIRED: "expired",
  // The feature flag is off. It answers 404 exactly as a build without the
  // feature does (ADR 0045), and the reader is told the link is not valid —
  // which is the true statement available to them.
  TICKET_QUESTIONS_UNAVAILABLE: "invalid",
  // No signing key configured: a deployment fault, not the holder's, so it says
  // "try again later" rather than blaming the link they were given.
  ANSWER_LINK_UNAVAILABLE: "unavailable",
} as const;

export type AnswerLinkFailure = (typeof ANSWER_LINK_FAILURE_KEYS)[keyof typeof ANSWER_LINK_FAILURE_KEYS];

/**
 * Which copy a failed open gets, from the API's error code.
 *
 * A code this app has not heard of falls through to "invalid", which is the
 * honest floor: the reader could not open their link, and the page has nothing
 * truer to tell them than that.
 */
export function answerLinkFailure(code: string | undefined): AnswerLinkFailure {
  if (code && code in ANSWER_LINK_FAILURE_KEYS) {
    return ANSWER_LINK_FAILURE_KEYS[code as keyof typeof ANSWER_LINK_FAILURE_KEYS];
  }
  return "invalid";
}

/**
 * The Options a choice question may be answered with RIGHT NOW: the live ones,
 * plus any retired one this Answer already chose.
 *
 * The second half is the point. A retired Option has left every new list, so it
 * is not offered to somebody answering afresh — but one already chosen must stay
 * on the form, because posting the Answer back without it would silently drop
 * what somebody said. "Kept on the Tickets that chose it" has to survive a
 * correction, not only a read.
 */
export function selectableOptions(pair: QuestionAnswer): QuestionOption[] {
  const chosen = new Set((pair.answer?.options ?? []).map((option) => option.option_id));
  return pair.question.options.filter((option) => !option.retired || chosen.has(option.id));
}

/** The Option ids this Answer currently holds, in the order it holds them. */
export function chosenOptionIds(pair: QuestionAnswer): string[] {
  return (pair.answer?.options ?? []).map((option) => option.option_id);
}

/**
 * The value a text, number or date field starts at: whatever is stored, or empty.
 *
 * One function rather than a branch per kind at the call site, so the form and
 * the payload cannot drift about what "no answer yet" looks like for a kind.
 *
 * WHAT IS STORED MAY BE THE BUYER'S GUESS. The holder sees it and may overwrite
 * it, which is the whole point of the link — somebody cannot correct a guess
 * they cannot see.
 */
export function initialFieldValue(pair: QuestionAnswer): string {
  const answer = pair.answer;
  if (!answer) return "";
  return answer.text ?? answer.number ?? answer.date ?? "";
}

/** Whether a checkbox question starts ticked. Absent is not the same as false. */
export function initialChecked(pair: QuestionAnswer): boolean {
  return pair.answer?.checked === true;
}

/**
 * The body to PUT for one question, given what the form holds.
 *
 * IT NEVER GUESSES THE KIND FROM THE VALUE. The kind comes from the question,
 * which is the only thing that knows it: the API refuses a body of the wrong
 * shape rather than coercing it, so a form sending `text` to a number question
 * is refused, and rightly.
 *
 * Returns null when there is nothing to send. An empty field is not an Answer of
 * "" — the way to say "not said" is to have no Answer at all — and this page
 * offers no way to take one back, deliberately: whoever holds the link may
 * OVERWRITE what is there, and erasing somebody else's reply from an
 * unauthenticated page is not a power this link grants (ADR 0044 gives the
 * buyer and Event Staff the correction; neither of them lost it).
 */
export function answerBodyFor(
  question: Question,
  value: string,
  checked: boolean,
  optionIds: string[],
): AnswerBody | null {
  switch (question.kind) {
    case "short_text":
    case "long_text":
      return value.trim() === "" ? null : { text: value };
    case "number":
      return value.trim() === "" ? null : { number: value };
    case "date":
      return value.trim() === "" ? null : { date: value };
    case "checkbox":
      // Always sendable, `false` included: false IS an answer.
      return { checked };
    case "single_choice":
    case "multi_choice":
      return optionIds.length === 0 ? null : { option_ids: optionIds };
  }
}

/**
 * The one Option's id toggled into or out of a choice Answer's selection.
 *
 * single_choice REPLACES rather than accumulates — picking M when S was chosen
 * means M, not both — while multi_choice adds and removes. The kind decides,
 * because a shared "toggle" that accumulated for both would let a single_choice
 * form build a body the API must refuse.
 */
export function toggleOption(
  question: Question,
  current: string[],
  optionId: string,
): string[] {
  if (question.kind === "single_choice") {
    return current.includes(optionId) ? [] : [optionId];
  }
  return current.includes(optionId)
    ? current.filter((id) => id !== optionId)
    : [...current, optionId];
}

/**
 * The questions this page draws, in the order it draws them.
 *
 * RETIRED QUESTIONS ARE KEPT ONLY IF THIS TICKET ANSWERED THEM, and are drawn
 * read-only by the form. A retired question is one the Organization has stopped
 * asking, so offering it to a holder answering afresh would be collecting an
 * answer to a question nobody wants any more — while dropping one this Ticket
 * HAS answered would make the reply look as though it had been thrown away, and
 * would hide from the holder something recorded about them.
 */
export function visibleQuestions(view: AnswerLinkView): QuestionAnswer[] {
  return view.questions.filter((pair) => !pair.question.retired || pair.answer !== null);
}

/** Whether a question is drawn read-only: retired ones, kept only to be read. */
export function isReadOnly(pair: QuestionAnswer): boolean {
  return pair.question.retired;
}

File: backend/internal/catalog/service/answer_link.go
package service

import (
	"context"

	"github.com/peter/ticket_pos/backend/internal/catalog"
	"github.com/peter/ticket_pos/backend/internal/catalog/repository"
)

// The Answer Link's own surface: minting one, opening one, and answering
// through one (#312, ADR 0044).
//
// EVERYTHING IN THIS FILE IS UNAUTHENTICATED. There is no ActorContext anywhere
// below, and that is the decision rather than an oversight: no sign-in, no
// One-time Passcode, no Customer created and no session minted. ADR 0044 records
// why — requiring proof of identity would mean collecting an address from
// somebody who never came to this platform, which is the third-party collection
// problem the whole feature exists to avoid. A wrong Answer is cheap, and both
// the buyer and Event Staff can fix it.
//
// The authority is the signed token and its whole reach is ONE Ticket's Ticket
// Questions. What keeps that safe is not authentication but DISCLOSURE: see
// AnswerLinkView, which is the boundary this ticket is actually about.

// answerLinkPath is the Storefront route an Answer Link points at.
//
// A STOREFRONT URL AND NEVER AN API ONE, like every other link this platform
// mails or hands out: the holder must land on a page, and no browser may address
// the Go API directly (ADR 0008).
//
// It is written WITHOUT a Locale prefix. The Storefront's middleware redirects
// an unprefixed path into a language and carries the query string with it, which
// is how the Confirmation Link and the Consent Confirmation Link already work.
// That matters more here than it does for those two: this link is pasted into a
// group chat by hand, and baking the buyer's Locale into it would hand a
// Spanish-speaking holder the buyer's English page.
//
// It is deliberately NOT under /tickets, which is the Customer Area. Holding this
// link makes nobody a Customer, and an address implying otherwise would be the
// first step toward somebody adding a "your tickets" link to the page.
const answerLinkPath = "/answer"

// AnswerLinkView is one Ticket as the holder of its Answer Link may see it.
//
// THIS TYPE IS THE SECURITY BOUNDARY OF THIS FEATURE. An Answer Link is an
// unauthenticated URL that gets forwarded into group chats, and ADR 0044 is
// explicit that its safety rests entirely on disclosing nothing about the
// purchase. So this is a SEPARATE TYPE from TicketAnswersView rather than a
// subset of it, assembled from a SEPARATE repository read that does not fetch
// the forbidden columns at all — a field that was never selected is not there to
// be leaked by somebody adding a `json:` tag.
//
// WHAT IS ABSENT, AND WHY EACH ONE STAYS ABSENT:
//
//   - The buyer's name and email. The holder is not the buyer and is owed
//     nothing about them; often they are a friend, and sometimes they are six
//     friends in a group chat.
//   - The price. What somebody paid for a gift is between them and the person
//     they bought it for.
//   - The Tax ID. It is a national identity number in this market.
//   - The Sale Confirmation reference. It is the buyer's credential-adjacent
//     handle on the sale, quoted to support, and printed on the receipt.
//   - The Ticket Sale's id, and the ordinal. Both say something about the
//     purchase's SHAPE — "you are the third of four" tells a group chat how many
//     tickets were bought, and an id is a thing to try somewhere else.
//   - Every other Ticket of the Sale. One link, one Ticket.
//
// ADR 0044 names the failure mode directly: "Anyone widening what the page shows
// — adding the buyer's name 'for context', or the confirmation reference 'to
// help support' — is reversing this decision." There is an integration test that
// asserts each of the above is absent from the raw response body, and it exists
// so that widening this struct is a red test rather than a code review somebody
// might wave through.
//
// NOR IS THERE A TICKET ID. The form posts back with the TOKEN, so the page
// never needs one, and not sending it means the response carries no identifier
// at all that could be tried against another surface.
type AnswerLinkView struct {
	// EventName is one of the three things the page shows. The Event is public —
	// it has a Storefront page anybody can read — so naming it discloses nothing
	// that was not already published.
	EventName string `json:"event_name"`
	// TicketTypeName is the second. Also public, for the same reason: it is a
	// row on that same Event page, with its price beside it. What is NOT here is
	// the price this particular buyer paid, which a Promotion or an Affiliate
	// Link may have made different from the published one.
	TicketTypeName string `json:"ticket_type_name"`
	// Questions is the third: this Ticket Type's Ticket Questions with whatever
	// this Ticket has already said. Labels read AS COINED in every Locale, like
	// a Custom Tag — only the page's chrome follows the reader's Locale.
	//
	// The Answers are shown because THE HOLDER IS ENTITLED TO SEE WHAT WAS SAID
	// ABOUT THEM. An Answer given here replaces one the buyer guessed at
	// checkout, and somebody cannot correct a guess they cannot see. This is
	// data about the holder, not about the purchase.
	Questions []TicketQuestionAnswerView `json:"questions"`
}

// AnswerLinkURL returns the Answer Link for one Ticket.
//
// This is what #315's buyer-facing page copies to a clipboard. It is a pure
// function of the Ticket id and the signing key: no database read, no Event
// lookup, and no row written anywhere — a stateless token needs no table, which
// is the reason this ticket has no migration.
//
// It does NOT check that the Ticket exists, or that its window is open. Both are
// read at the moment the link is OPENED, which is the only moment at which the
// answer is still true; a check here would be a check against a state the link
// will outlive.
func (s *Service) AnswerLinkURL(ticketID string) (string, error) {
	token, ok := s.answerLinks.Sign(ticketID)
	if !ok {
		return "", catalog.ErrAnswerLinkUnavailable()
	}
	return s.answerLinkBaseURL + answerLinkPath + "?token=" + token, nil
}

// OpenAnswerLink turns a token into the one Ticket's questions it opens.
//
// THE ORDER OF THE REFUSALS IS THE DESIGN. The feature flag is read before the
// token is even parsed, so a dark build answers exactly as a build that never
// had the feature (ADR 0045). Then the signature, so a forged or truncated token
// never reaches a database lookup. Then the Ticket, then the window.
//
// EVERY REFUSAL BUT ONE IS THE SAME REFUSAL. A bad signature, a Ticket that no
// longer exists and a REVERSED Sale are all ANSWER_LINK_INVALID, because telling
// them apart would let whoever holds the link learn something about the
// purchase — that it existed, that it was refunded — which is the one thing this
// page must never do. Only "the Event has started" is told apart, and only
// because an Event's start is already published.
func (s *Service) OpenAnswerLink(ctx context.Context, token string) (*AnswerLinkView, error) {
	ticket, err := s.answerLinkTicket(ctx, token)
	if err != nil {
		return nil, err
	}
	return s.answerLinkView(ctx, ticket)
}

// AnswerByAnswerLink writes one Ticket's Answer to one of its Ticket Questions,
// on the authority of the token alone.
//
// AN ANSWER GIVEN HERE REPLACES ONE THE BUYER GAVE AT CHECKOUT. There is nothing
// special about that in the code and there should not be: an Answer belongs to
// the TICKET, so this is the same upsert through the same (ticket, question)
// UNIQUE that Event Staff go through, and "replaces" is what one row per pair
// means. Nothing records that the holder rather than the buyer wrote it —
// migration 073 has no author column on purpose.
//
// It reuses AnswerTicketQuestion's whole body through answerTicketQuestion,
// which is where catalog.ParseAnswer, the retired-question rule, the Option
// resolution and the "a re-submission is not a change" guard all live. Three
// surfaces answering the same question — is this a thing this Ticket Question
// can be answered with — would be three ways for a NUMERIC column to end up
// holding something that is not a number.
func (s *Service) AnswerByAnswerLink(
	ctx context.Context,
	token, questionID string,
	input AnswerInput,
) (*AnswerLinkView, error) {
	ticket, err := s.answerLinkTicket(ctx, token)
	if err != nil {
		return nil, err
	}

	if err := s.answerTicketQuestion(ctx, ticket.ID, ticket.TicketTypeID, questionID, input); err != nil {
		return nil, err
	}
	return s.answerLinkView(ctx, ticket)
}

// answerLinkTicket is the single gate every route in this file passes through:
// flag, signature, Ticket, window.
//
// ONE GATE AND NOT TWO, because the read and the write must refuse identically.
// If opening a link on a reversed Sale were invalid but answering through one
// were merely refused, the write would confirm the Ticket exists to somebody the
// read had just told nothing.
func (s *Service) answerLinkTicket(ctx context.Context, token string) (*repository.AnswerLinkTicket, error) {
	// The flag first, before anything is parsed or read. While the feature is
	// dark this address must be indistinguishable from one that was never
	// routed — same status, same code, and no read whose timing could differ
	// between a real Ticket and an invented one (ADR 0045).
	if !s.ticketQuestionsEnabled {
		return nil, catalog.ErrTicketQuestionsUnavailable()
	}
	if !s.answerLinks.Configured() {
		return nil, catalog.ErrAnswerLinkUnavailable()
	}

	ticketID, ok := s.answerLinks.Parse(token)
	if !ok {
		return nil, catalog.ErrAnswerLinkInvalid()
	}

	ticket, err := s.repo.GetAnswerLinkTicket(ctx, ticketID)
	if err != nil {
		return nil, err
	}
	if ticket == nil {
		// Signed by us and naming nothing — a Ticket whose Event was deleted
		// while it was a draft, or whose Sale Import was undone. Reported as an
		// invalid link and never as a missing Ticket: the holder is not entitled
		// to learn which of the two it was.
		return nil, catalog.ErrAnswerLinkInvalid()
	}

	// THE ANSWER LINK STOPS OPENING ONCE A HOLDER HAS ACCEPTED (#325, ADR 0046).
	//
	// This is the one thing accepting takes away, and it is what makes accepting
	// mean anything. Both doors stand open while a Ticket is merely `assigned`,
	// so a mistyped address bricks nothing and an ignored mail degrades to
	// exactly ADR 0044's behaviour. The moment somebody PROVES the address, this
	// door is retired in their favour: a link still sitting in a group chat, or
	// forwarded on to six people, cannot overwrite the Holder's own answer about
	// their own body as a joke.
	//
	// IT IS ANSWER_LINK_INVALID AND NOT A NEW CODE. Whoever holds this link is
	// entitled to learn that it no longer works and nothing else — that somebody
	// else has claimed this Ticket is a fact about a person, and this page has
	// never disclosed one.
	//
	// THE BUYER AND EVENT STAFF KEEP THEIR ROUTES. ADR 0044's three-party rule
	// holds: this closes the unauthenticated door, not the two accountable ones,
	// so a wrong Answer stays fixable.
	if ticket.AcceptedAt.Valid {
		return nil, catalog.ErrAnswerLinkInvalid()
	}

	switch catalog.AnswerWindow(ticket.SaleStatus, nullTimeOrNil(ticket.EventStartsAt), s.now()) {
	case catalog.AnswerRefusedSaleReversed:
		// NOT ErrTicketSaleReversed, which is what Event Staff hear. Staff are
		// entitled to know a sale was reversed and to keep reading its Answers;
		// a group chat is entitled to neither. See catalog.ErrAnswerLinkInvalid.
		return nil, catalog.ErrAnswerLinkInvalid()
	case catalog.AnswerRefusedEventStarted:
		return nil, catalog.ErrAnswerLinkExpired()
	}
	return ticket, nil
}

// answerLinkView assembles the payload from the Ticket Type's questions and this
// Ticket's Answers.
//
// It reuses ticketQuestionViews and the Answer read, which is why retired
// questions and retired Options arrive here too. That is right for the same
// reason it is right on the staff surface: an Answer against a retired Option
// has to keep reading, and a question whose Answer vanished from the page would
// look like the holder's reply had been thrown away.
func (s *Service) answerLinkView(ctx context.Context, ticket *repository.AnswerLinkTicket) (*AnswerLinkView, error) {
	questions, err := s.ticketQuestionViews(ctx, ticket.TicketTypeID)
	if err != nil {
		return nil, err
	}
	answers, err := s.repo.ListTicketAnswers(ctx, []string{ticket.ID})
	if err != nil {
		return nil, err
	}
	byQuestion := make(map[string]int, len(answers))
	for i, answer := range answers {
		byQuestion[answer.TicketQuestionID] = i
	}

	view := AnswerLinkView{
		EventName:      ticket.EventName,
		TicketTypeName: ticket.TicketTypeName,
		Questions:      make([]TicketQuestionAnswerView, 0, len(questions)),
	}
	for _, question := range questions {
		pair := TicketQuestionAnswerView{Question: question}
		if i, found := byQuestion[question.ID]; found {
			pair.Answer = toAnswerView(answers[i])
		}
		view.Questions = append(view.Questions, pair)
	}
	return &view, nil
}
